File: src/main/java/com/chatdemo/memory/HistoryAwareChatMemoryStore.java
package com.chatdemo.memory;

import com.chatdemo.model.ConversationMessage;
import com.chatdemo.repository.ConversationRepository;
import dev.langchain4j.data.message.ChatMessage;
import dev.langchain4j.data.message.SystemMessage;
import dev.langchain4j.store.memory.chat.ChatMemoryStore;

import java.util.ArrayList;
import java.util.List;

/**
 * ChatMemoryStore adapter that bridges LangChain4j's memory system to our ConversationRepository.
 * 
 * This store:
 * - Returns summary + active messages when getMessages() is called
 * - Detects newly added messages and persists them
 * - Does NOT handle archiving directly (that's done by SummarizingTokenWindowChatMemory)
 */
public class HistoryAwareChatMemoryStore implements ChatMemoryStore {
    
    private final ConversationRepository repository;
    
    public HistoryAwareChatMemoryStore(ConversationRepository repository) {
        this.repository = repository;
    }
    
    @Override
    public List<ChatMessage> getMessages(Object memoryId) {
        String conversationId = memoryId.toString();
        
        List<ChatMessage> result = new ArrayList<>();
        
        // Add summary as SystemMessage if present
        String summary = repository.getSummary(conversationId);
        if (summary != null && !summary.isEmpty()) {
            result.add(SystemMessage.from("Summary of earlier conversation:\n" + summary));
        }
        
        // Add active (non-archived) messages
        result.addAll(repository.getActiveMessages(conversationId));
        
        return result;
    }
    
    @Override
    public void updateMessages(Object memoryId, List<ChatMessage> messages) {
        String conversationId = memoryId.toString();
        
        // Get current active messages from repository
        List<ChatMessage> currentMessages = repository.getActiveMessages(conversationId);
        
        // Find messages in the incoming list that are NOT the summary SystemMessage
        // and are not already in the repository
        for (ChatMessage message : messages) {
            // Skip summary system messages (they start with "Summary of earlier conversation")
            if (message instanceof SystemMessage systemMessage) {
                if (systemMessage.text().startsWith("Summary of earlier conversation")) {
                    continue;
                }
            }
            
            // Check if this message already exists in the repository
            if (!containsMessage(currentMessages, message)) {
                repository.addMessage(conversationId, message);
            }
        }
    }
    
    @Override
    public void deleteMessages(Object memoryId) {
        repository.clear(memoryId.toString());
    }
    
    /**
     * Archives messages and updates the summary.
     * Called by SummarizingTokenWindowChatMemory when summarization occurs.
     */
    public void archiveMessages(String conversationId, List<String> messageIds, String summary) {
        repository.archiveMessages(conversationId, messageIds, summary);
    }
    
    /**
     * Get message IDs for active messages (used by summarizing memory to know what to archive).
     */
    public List<String> getActiveMessageIds(String conversationId) {
        return repository.getFullHistory(conversationId).stream()
            .filter(msg -> !msg.isArchived())
            .map(ConversationMessage::getId)
            .toList();
    }
    
    /**
     * Get full history for UI display.
     */
    public List<ConversationMessage> getFullHistory(String conversationId) {
        return repository.getFullHistory(conversationId);
    }
    
    private boolean containsMessage(List<ChatMessage> messages, ChatMessage target) {
        for (ChatMessage msg : messages) {
            if (messagesEqual(msg, target)) {
                return true;
            }
        }
        return false;
    }
    
    private boolean messagesEqual(ChatMessage a, ChatMessage b) {
        if (a.getClass() != b.getClass()) {
            return false;
        }
        return a.equals(b);
    }
}

File: src/main/scala/com/chatdemo/backend/memory/HistoryAwareChatMemoryStore.scala
package com.chatdemo.backend.memory

import com.chatdemo.backend.repository.ConversationRepository
import com.chatdemo.common.model.{ConversationMessage, UserContext}
import dev.langchain4j.data.message.{AiMessage, ChatMessage, SystemMessage, UserMessage}
import dev.langchain4j.store.memory.chat.ChatMemoryStore

import java.util.{List => JList}
import scala.collection.mutable.ArrayBuffer
import scala.jdk.CollectionConverters.*

/**
 * ChatMemoryStore adapter that bridges LangChain4j's memory system to our ConversationRepository.
 *
 * This store:
 * - Returns summary + active messages when getMessages() is called
 * - Detects newly added messages and persists them
 * - Does NOT handle archiving directly (that's done by SummarizingTokenWindowChatMemory)
 */
class HistoryAwareChatMemoryStore(
  private val repository: ConversationRepository,
  private val userContext: UserContext
) extends ChatMemoryStore {

  private val persistedMessages: ArrayBuffer[ChatMessage] = ArrayBuffer.empty

  override def getMessages(memoryId: AnyRef): JList[ChatMessage] = {
    val conversationId = memoryId.toString
    val result = new java.util.ArrayList[ChatMessage]()
    persistedMessages.clear()

    // Add summary as SystemMessage if present
    val summary = repository.getSummary(userContext, conversationId)
    if (summary != null && summary.nonEmpty) {
      result.add(SystemMessage.from("Summary of earlier conversation:\n" + summary))
    }

    // Add active (non-archived) messages.
    // Skip persisted system prompts so stale prompt versions do not affect future turns.
    for (msg <- repository.getActiveMessages(userContext, conversationId)) {
      msg match {
        case _: SystemMessage => // skip persisted system prompts
        case _ =>
          if (!isNoiseMessage(msg)) {
            result.add(msg)
            persistedMessages.append(msg)
          }
      }
    }

    result
  }

  override def updateMessages(memoryId: AnyRef, messages: JList[ChatMessage]): Unit = {
    val conversationId = memoryId.toString
    val currentMessages = persistedMessages

    for (message <- messages.asScala) {
      // Skip summary system messages
      message match {
        case sm: SystemMessage if sm.text().startsWith("Summary of earlier conversation") =>
          // skip
        case _: SystemMessage =>
          // Do not persist runtime system prompts in conversation history.
          // The current system prompt is injected per-request.
        case _ =>
          if (!isNoiseMessage(message) && !containsMessage(currentMessages, message)) {
            repository.addMessage(userContext, conversationId, message)
            currentMessages.append(message)
          }
      }
    }
  }

  override def deleteMessages(memoryId: AnyRef): Unit = {
    repository.clear(userContext, memoryId.toString)
  }

  /** Archives messages and updates the summary. */
  def archiveMessages(conversationId: String, messageIds: List[String], summary: String): Unit = {
    repository.archiveMessages(userContext, conversationId, messageIds, summary)
  }

  /** Get message IDs for active messages. */
  def getActiveMessageIds(conversationId: String): List[String] = {
    repository.getFullHistory(userContext, conversationId)
      .filter(!_.archived)
      .map(_.id)
  }

  /** Get full history for UI display. */
  def getFullHistory(conversationId: String): List[ConversationMessage] = {
    repository.getFullHistory(userContext, conversationId)
  }

  private def containsMessage(messages: scala.collection.Iterable[ChatMessage], target: ChatMessage): Boolean = {
    messages.exists(msg => messagesEqual(msg, target))
  }

  private def isNoiseMessage(message: ChatMessage): Boolean = {
    def containsToolResultMarker(text: String): Boolean = {
      text != null && text.contains("ToolExecutionResultMessage")
    }

    message match {
      case ai: AiMessage =>
        val text = ai.text()
        text == null || text.isBlank || containsToolResultMarker(text)
      case user: UserMessage =>
        val text = user.singleText()
        text == null || text.isBlank || containsToolResultMarker(text)
      case other =>
        val text = other.toString
        text == null || text.isBlank || containsToolResultMarker(text)
    }
  }

  private def messagesEqual(a: ChatMessage, b: ChatMessage): Boolean = {
    if (a.getClass != b.getClass) {
      false
    } else {
      a == b
    }
  }
}
